core/src/main/java/com/linecorp/armeria/internal/common/HttpObjectEncoder.java

```diff
@@ -45,19 +45,21 @@ protected EventLoop eventLoop() {
     /**
      * Writes an {@link HttpHeaders}.
      */
-    public final ChannelFuture writeHeaders(int id, int streamId, HttpHeaders headers, boolean endStream) {
-
+    public final ChannelFuture writeHeaders(int id, int streamId, HttpHeaders headers, boolean endStream,
+                                            HttpHeaders additionalHeaders, HttpHeaders additionalTrailers) {
```

Comment:
Yes, I think it's good idea. I will update this soon :grinning:

Reply:
It seems weird to me that this `writeHeaders` method has both `additionalHeaders` and `additionalTrailers` which cannot be used at the same time.
How about splitting `writeHeaders` to `writeHeaders` and `writeTrailers`?

Reply:
I think `additionalHeaders` and `additionalTrailers` can be used at the same time when request headers end with additional trailers. (Do I understand correctly? :thinking: )

It seems that it may happens here now.
https://github.com/line/armeria/blob/48f413285c15b269d4ccbc8ccfb5b5e5cfbb799f/core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java#L174

Do you think we need all possible overrides for this method?
``` java
public final ChannelFuture writeHeaders(int id, int streamId, HttpHeaders headers, boolean endStream)
public final ChannelFuture writeHeaders(int id, int streamId, HttpHeaders headers, boolean endStream, HttpHeaders additionalHeaders)
public final ChannelFuture writeTrailers(int id, int streamId, HttpHeaders headers, boolean endStream, HttpHeaders additionalTrailers)
public final ChannelFuture writeHeaders(int id, int streamId, HttpHeaders headers, boolean endStream, HttpHeaders additionalHeaders, HttpHeaders additionalTrailers)
```

Reply:
> It seems that it may happens here now.

I think it's a bug so I fixed it. 😄
#2544
After this is merged, I think we can split the method to `writeHeaders` and `writeTrailers`. Sorry about the bug. 😉

Reply:
#2544 has been fixed. @jyblue, could you rebase and squash into a single commit, since the history is getting pretty large?

Reply:
I squashed and rebase commit. I will also split `writeHeaders` methods soon. :bowing_man: